Eulerian Tours editorial (#3995)
cpinitiative/usaco-guide: content/6_Advanced/Eulerian_Tours.mdx

## Implementation

### Mail Delivery

First let's define what an **Eulerian** path is

> An Eulerian path is a path that goes through every edge once.

Similarly, an **Eulerian cycle** is an Eulerian path that starts and ends
with the same node.

> An important condition is that a graph can be an Eulerian graph if and only if every node has an even degree.

Reviewer: i thought a graph w/ 2 odd degree nodes also was possible?

Author: Nope. It's necessary for an Eulerian path that all noded have even degree.
https://en.wikipedia.org/wiki/Eulerian_path

Now, to properly find the Eulerian cycle we run a slightly modified DFS.

> The DFS goes through only unvisited edges and the same edge can be processe multiple times throuoght the DFs, so we remove it from the graph at the first visit.

Reviewer 2: ```suggestion
The DFS goes through only unvisited edges and the same edge can be processed multiple times throughout the DFS, so we remove it from the graph at the first visit.
```

Author: Alright

The algorithm described is [Hierholzer's Algorithm](https://en.wikipedia.org/wiki/Eulerian_path).

> 

Reviewer 2: ```suggestion
## Implementation
```

add time complexity as well

Author: Ok

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

int n, m;
vector<vector<pair<int, int>>> g;

> vector<int> degree, path;

Reviewer 2: does degree need to be global?

Author: OK, it's local now

vector<bool> seen;

void dfs(int node) {
	while (!g[node].empty()) {
		auto [son, idx] = g[node].back();
		g[node].pop_back();
		if (seen[idx]) { continue; }
		seen[idx] = true;
		dfs(son);
	}
	path.push_back(node);
}

int main() {
	cin >> n >> m;

	g.resize(n);
	degree.resize(n);
	seen.resize(m);

	for (int i = 0; i < m; i++) {
		int x, y;
		cin >> x >> y;
		x--, y--;
		g[x].push_back({y, i});

> 		g[y].push_back({x, i});

Reviewer 2: emplace_back ? optional

Author: Ok changed

		degree[x]++;
		degree[y]++;
	}

	for (int node = 0; node < n; node++) {
		if (degree[node] % 2) {
			cout << "IMPOSSIBLE";
			return 0;
		}
	}

	dfs(0);

	if (path.size() != m + 1) {
		cout << "IMPOSSIBLE";
	} else {
		for (auto node : path) { cout << node + 1 << ' '; }
	}
}
```

</CPPSection>
</LanguageSection>

### Teleporters

> The condition of existance for an eulerian path in a directed graph is: At most one node has $out_i - in_i=1$ and at most one node has $in_i - out_i=1$.

Reviewer 2: can u explain why?

Author: Ok  I'll explain

<LanguageSection>
<CPPSection>
```cpp
#include <bits/stdc++.h>
using namespace std;

int n, m;
vector<vector<int>> g;
vector<int> in, out, path;

void dfs(int node) {
	while (!g[node].empty()) {
		int son = g[node].back();
		g[node].pop_back();
		dfs(son);
	}
	path.push_back(node);
}

int main() {
	cin >> n >> m;

	g.resize(n + 1);
	in.resize(n + 1);
	out.resize(n + 1);

	for (int i = 0; i < m; i++) {
		int x, y;
		cin >> x >> y;
		g[x].push_back(y);
		out[x]++;
		in[y]++;
	}

	bool flag = true;
	for (int node = 2; node < n && flag; node++) {
		if (in[node] != out[node]) { flag = false; }
	}
	if (out[1] != in[1] + 1 || out[n] != in[n] - 1 || !flag) {
		cout << "IMPOSSIBLE";
		return 0;
	}

	dfs(1);

	reverse(path.begin(), path.end());
	if (path.size() != m + 1 || path.back() != n) {
		cout << "IMPOSSIBLE";
	} else {
		for (auto node : path) { cout << node << ' '; }
	}

> 	return 0;

Reviewer 2: ```suggestion
```

Author: Done
